src/rsz/src/Resizer.cc

```diff
@@ -2701,38 +2718,98 @@ Resizer::journalSwapPins(Instance *inst, LibertyPort *port1,
+  if (new_journal_) {
+    journal_->swapPins(inst, port1, port2);
+  }
+  else {
+    swapped_pins_[inst] = std::make_tuple(port1, port2);
+  }
 }
 
 void
 Resizer::journalInstReplaceCellBefore(Instance *inst)
 {
-  LibertyCell *lib_cell = network_->libertyCell(inst);
-  debugPrint(logger_, RSZ, "journal", 1, "journal replace {} ({})",
-             network_->pathName(inst),
-             lib_cell->name());
-  // Do not clobber an existing checkpoint cell.
-  if (!resized_inst_map_.hasKey(inst))
-    resized_inst_map_[inst] = lib_cell;
+  if (new_journal_) {
+    journal_->instReplaceCellBefore(inst);
+  }
+  else {
+    // Do not clobber an existing checkpoint cell.
+    LibertyCell* lib_cell = network_->libertyCell(inst);
+    debugPrint(logger_, RSZ, "journal", 1, "journal replace {} ({})", network_->pathName(inst), lib_cell->name());
+    if (!resized_inst_map_.hasKey(inst)) {
+      resized_inst_map_[inst] = lib_cell;
+    }
+  }
 }
 
 void
 Resizer::journalMakeBuffer(Instance *buffer)
 {
-  debugPrint(logger_, RSZ, "journal", 1, "journal make_buffer {}",
-             network_->pathName(buffer));
-  inserted_buffers_.emplace_back(buffer);
-  inserted_buffer_set_.insert(buffer);
+  if (new_journal_) {
+    journal_->makeBuffer(buffer);
+  }
+  else {
+    debugPrint(logger_, RSZ, "journal", 1, "journal make_buffer {}", network_->pathName(buffer));
+    inserted_buffers_.emplace_back(buffer);
+    inserted_buffer_set_.insert(buffer);
+  }
+}
+
+int Resizer::undoGateCloning(Instance *original_inst, Instance *cloned_inst) {
+  debugPrint(logger_, RSZ, "journal", 1, "journal unclone {} ({}) -> {} ({})",
+             network_->pathName(original_inst),
+             network_->libertyCell(original_inst)->name(),
+             network_->pathName(cloned_inst),
+             network_->libertyCell(cloned_inst)->name());
+
+  const Pin *original_output_pin = nullptr;
+  std::vector<const Pin *> clone_pins;
+  getPins(cloned_inst, clone_pins);
+  std::vector<const Pin *> original_pins;
+  getPins(original_inst, original_pins);
+  for (auto &pin: original_pins) {
+    if (network_->direction(pin)->isOutput()) {
+      original_output_pin = pin;
+      break;
+    }
+  }
+  Net *original_out_net = network_->net(original_output_pin);
+  // Net* clone_out_net = nullptr;
+
+  for (auto &pin: clone_pins) {
+    // Disconnect all pins from the new net. Also store the output net
+    // if (network_->direction(pin)->isOutput()) {
+    //  clone_out_net = network_->net(pin);
+    //}
+    sta_->disconnectPin(const_cast<Pin *>(pin));
+    // Connect them to the original nets if they are inputs
+    if (network_->direction(pin)->isInput()) {
+      Instance *inst = network_->instance(pin);
+      auto term_port = network_->port(pin);
+      sta_->connectPin(inst, term_port, original_out_net);
+    }
+  }
+  // Final cleanup
+  // sta_->deleteNet(clone_out_net);
+  sta_->deleteInstance(cloned_inst);
+  sta_->graphDelayCalc()->delaysInvalid();
```

Comment:
Shouldn't sta do this automatically?

Reply:
I am assuming you mean the delaysInvalid call. Fixed.